tftp_test.go

```diff
@@ -0,0 +1,211 @@
+package ipxe
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pin/tftp"
+	"github.com/tinkerbell/boots-ipxe/binary"
+	"go.opentelemetry.io/otel/trace"
+	"inet.af/netaddr"
+)
+
+type fakeReaderFrom struct {
+	addr    net.UDPAddr
+	content []byte
+	err     error
+}
+
+func (f *fakeReaderFrom) ReadFrom(r io.Reader) (n int64, err error) {
+	if f.err != nil {
+		return 0, f.err
+	}
+	nInt, err := r.Read(f.content)
+	return int64(nInt), err
+}
+
+func (f *fakeReaderFrom) SetSize(_ int64) {}
+
+func (f *fakeReaderFrom) RemoteAddr() net.UDPAddr {
+	return f.addr
+}
+
+func (f *fakeReaderFrom) WriteTo(_ io.Writer) (n int64, err error) {
+	return 0, nil
+}
+
+func TestListenAndServeTFTP(t *testing.T) {
+	ht := &HandleTFTP{Log: logr.Discard()}
+	srv := tftp.NewServer(ht.ReadHandler, ht.WriteHandler)
+	type args struct {
+		ctx  context.Context
+		addr netaddr.IPPort
+		h    *tftp.Server
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr interface{}
+	}{
+		{
+			name: "fail",
+			args: args{
+				ctx:  context.Background(),
+				addr: netaddr.IPPortFrom(netaddr.IPv4(127, 0, 0, 1), 80),
+				h:    nil,
+			},
+			wantErr: &net.OpError{},
+		},
+		{
+			name: "success",
+			args: args{
+				ctx:  context.Background(),
+				addr: netaddr.IPPortFrom(netaddr.IPv4(127, 0, 0, 1), 9999),
+				h:    srv,
+			},
+			wantErr: interface{}(nil),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			errChan := make(chan error, 1)
+			go func() {
+				errChan <- ListenAndServeTFTP(tt.args.ctx, tt.args.addr, tt.args.h)
+			}()
+
+			if tt.args.h != nil {
+				tt.args.ctx.Done()
+				time.Sleep(time.Second)
```

Comment:
Does `<- tt.args.ctx.Done()` not work instead of the sleep? I know it didn't work for me in a different project's test but I attributed that to `grpc`'s async ness. I think http shouldn't have the same issue.

Reply:
in this instance, when i replace `time.Sleep(time.Second)` with `<-tt.args.ctx.Done()` will just sit forever or until the `go test` timeout. Also, this is not HTTP.

Reply:
> in this instance, when i replace `time.Sleep(time.Second)` with `<-tt.args.ctx.Done()` will just sit forever or until the `go test` timeout.

Ah yeah I see now. I also tried some things locally and none of it works because there's no way to know when ListenAndServeTFTP is ready to actually serve. In that case what does `tt.args.ctx.Done()` even provide here? Done doesn't do any clean up it just returns a channel (according to docs) so it seems to me it can go away all together right?

> Also, this is not HTTP.

:man_facepalming: right!

Reply:
sure thats fair. the context.Done() isnt needed. will update.